File: myExpenses/src/main/java/org/totschnig/myexpenses/sync/json/ChangeSet.java
package org.totschnig.myexpenses.sync.json;


import org.totschnig.myexpenses.sync.SequenceNumber;

import java.util.ArrayList;
import java.util.List;

public class ChangeSet {

  public final SequenceNumber sequenceNumber;
  public final List<TransactionChange> changes;

  private ChangeSet(SequenceNumber sequenceNumber, List<TransactionChange> changes) {
    this.sequenceNumber = sequenceNumber;
    this.changes = changes;
  }

  public static ChangeSet create(SequenceNumber sequenceNumber, List<TransactionChange> changes) {
    return new ChangeSet(sequenceNumber, changes);
  }


  public static ChangeSet empty(SequenceNumber sequenceNumber) {
    return create(sequenceNumber, new ArrayList<>());
  }

  public static ChangeSet merge(ChangeSet changeset1, ChangeSet changeset2) {
    if (changeset1 == null || changeset2 == null) {
      return null;
    }
    List<TransactionChange> changes = new ArrayList<>();
    changes.addAll(changeset1.changes);
    changes.addAll(changeset2.changes);
    SequenceNumber max = SequenceNumber.max(changeset1.sequenceNumber, changeset2.sequenceNumber);
    return new ChangeSet(max, changes);
  }
}

File: myExpenses/src/main/java/org/totschnig/myexpenses/sync/json/ChangeSet.kt
package org.totschnig.myexpenses.sync.json

import org.totschnig.myexpenses.sync.SequenceNumber
import org.totschnig.myexpenses.sync.SequenceNumber.Companion.max
import org.totschnig.myexpenses.sync.json.TransactionChange
import org.totschnig.myexpenses.sync.json.ChangeSet
import java.util.ArrayList

class ChangeSet private constructor(
    val sequenceNumber: SequenceNumber,
    val changes: List<TransactionChange>
) {
    companion object {
        fun create(sequenceNumber: SequenceNumber, changes: List<TransactionChange>): ChangeSet {
            return ChangeSet(sequenceNumber, changes)
        }

        fun empty(sequenceNumber: SequenceNumber): ChangeSet {
            return create(sequenceNumber, ArrayList())
        }

        fun merge(changeSet1: ChangeSet?, changeSet2: ChangeSet?): ChangeSet? {
            if (changeSet1 == null || changeSet2 == null) {
                return null
            }
            val changes: MutableList<TransactionChange> = ArrayList()
            changes.addAll(changeSet1.changes)
            changes.addAll(changeSet2.changes)
            val max = max(changeSet1.sequenceNumber, changeSet2.sequenceNumber)
            return ChangeSet(max, changes)
        }
    }
}